ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsController.java

```diff
@@ -140,104 +47,65 @@ public synchronized void onPreviewOrderChanged(StepMovedEvent event) {
-     * Made to be called in {@link PreviewsController#onSocketPreviewChanged}
-     *
-     * @param socket An output socket associated with a step (NOT a source)
-     * @return The correct index in the list of displayed previews for the given <code>socket</code>
-     * @see PreviewsController#onSocketPreviewChanged(SocketPreviewChangedEvent)
-     */
-    private int getIndexInPreviewsOfAStepSocket(OutputSocket<?> socket) {
-        int numbOfSourcePreviews = getNumbOfSourcePreviews();//Count how many *source* previews (not *step* previews) are currently displayed
-
-        final Step socketStep = socket.getStep().get();//The pipeline step associated with the socket whose preview has changed
-        int indexOfStep = this.pipeline.getSteps().indexOf(socketStep); //The index of the step that has the socket in the pipeline
-
-        //Start at the first non-source socket in the list of previewed sockets
-        long indexInPreviews =
-                // The socket at this index in the list of displayed sockets has an index in the pipeline less than the socket passed in as "socket"
-                this.previewedSockets.stream().filter(outSocket -> outSocket.getStep().isPresent() && this.pipeline.getSteps().indexOf(outSocket.getStep().get()) < indexOfStep).count();
-        return IntMath.checkedAdd(Math.toIntExact(indexInPreviews), numbOfSourcePreviews);
-    }
-
-    /**
-     * Counts how many source previews (NOT step previews) are currently displayed.
-     * Called in  {@link PreviewsController#getIndexInPreviewsOfAStepSocket} and {@link PreviewsController#onPreviewOrderChanged(StepMovedEvent)}
-     *
-     * @return The number of source (NOT step) previews that are currently displayed
-     * @see PreviewsController#getIndexInPreviewsOfAStepSocket(OutputSocket)
-     * @see PreviewsController#onPreviewOrderChanged(StepMovedEvent)
-     */
-    private int getNumbOfSourcePreviews() {
-        //Start at the beginning of the list.
-        int numbOfSourcePreviews = 0;
-        while ((this.previewedSockets.size() > numbOfSourcePreviews) //While there are still previews to examine
-                && (!this.previewedSockets.get(numbOfSourcePreviews).getStep().isPresent())) { //If this is a source...
-            numbOfSourcePreviews++;
+        if (a.getSource().isPresent() && b.getSource().isPresent()) {
+            final Source sourceA = a.getSource().get(), sourceB = b.getSource().get();
+
+            if (sourceA == sourceB) {
+                // If both sockets are in the same source, order them based on which is first in the source
+                return Arrays.asList(sourceA.getOutputSockets()).stream()
+                        .filter(socket -> socket == a || socket == b)
+                        .findFirst().get() == a ? -1 : 1;
+            } else {
+                // If both sockets are from sources, order them based on the order of the sources in the pipeline
+                return pipeline.getSources().stream()
```

Comment:
There is no guarantee that the source will be added to the pipeline first. How do you know that this will run only after it has been added to the pipeline?